### ISSUE TITLE
7.0.0.5 - Square Baler slows down to eject bales

### ISSUE THREAD

**Issue by dansiegel1109** [Courseplay/Courseplay_FS22#143]
**Steps to reproduce**
Run a course with a Big450M 10m mower
Save Course
Run a course with a tractor attached to the Krone BigPack Square Baler

**Actual behavior**
Baling works at 17kmph is max working speed is set to 17kmph
However, when the baler is full, the vehicle stops to eject and then continues
The vehicle actually never reaches 0kmph, slowest it gets to is 3kmph and immediately resumes max speed of 17kmph

**Expected behavior**
The square baler does not need to stop to eject a bale.  I am assuming that CP believes it must stop, sets the speed to 0kmph, however prior to actually stopping, it realizes it is not full anymore and resumes max speed.  For square balers  (and the round fast baler), I think the best thing for CP is to allow it to continue for 2 seconds after the bale should be ejected.  If after 2 seconds, it realizes it doesn't need to, then all is well and let it continue filling next bale.  For traditional round balers which must stop, it will obviously run over product which needs to be picked up.  At this time, it can reverse those few meters - the round baler should still be in the process of ejecting, so there should be no time lost (this is how it works when you do it manually, for the best implement efficiency).

This suggested solution may have consequences with Fast Baler auto ejection for the bales.  I have not tested the fast round baler.  I assume CP will always enable auto ejection when possible or available - an advanced feature of this would be to not eject during a turn, but only on a straight or relatively straight line (a course with an extremely long curved course is I suppose possible).

**Screenshots**
n/a

**Log**
The problem is trivial enough where I do not believe logs are required. However, please let me know if you need them.

**Map**
Elm Creek - Field 70 - load new game, sleep 2 months prior to test to let grass grow

**Vehicles/ implements involved**
Krone BigPack Square Baler - I assume the same problem to happen with any other square baler or round fast baler

**Relevant script mods**
none

**Comment by Tensuko** [comment]
> The problem is trivial enough where I do not believe logs are required. However, please let me know if you need them.

But it still gives us Informations!

**Comment by dansiegel1109** [comment]
Fair enough.  I will provide the debug channel info, along with tests of the square, fast round, and traditional (which must stop) balers.  Danke!

**Comment by dansiegel1109** [comment]
Log file with several debug channels, including Bale Mode 7 

Several Balers tested, with one surprising result - from looking at the logs, I am assuming you should be able to differentiate between the different implements tested, however the below list is in the exact order I tested them:
1. Pottinger 125cm round Baler - it does need to stop to eject, and worked perfectly!
2. Vicon Fast Baler - surprise - this baler does not need to stop, however it did NOT slow down - perhaps due to it's ante-chamber
3. Krone BigPack Square Baler with BonSilage (official Modhub mod) - this baler does slow down when the crop level is at 100% - note, when it physically ejects the bale onto the ground, it does not slow down (it can carry 1 full bale while creating another)
4. Class Square Baler - same as Krone
5. Kuhn Square Baler - same as Krone


In a way, both the Vicon and all Square balers are "fast" balers - the Vicon however is implemented differently than the square balers, as the Vicon as an ante-chamber for creating the bale, and holds the previously made bale until you wish to eject it (or auto eject).  As well, you can see the previously created bale and its size (always at 100%) on the Vicon.  Square balers, the original fast balers, are more archaic...they do not officially have an ante-chamber.  When the next bale begins processing, the Square Baler holds the previously made bale, however it is not shown on the HUD.  The only way you can rapid eject this bale at the time of your choosing is if you turn off the baler - the Vicon is a superior design by Giants compared to the square baler, and they appear to be behaving differently in regards to this problem.

[log.zip](https://github.com/Courseplay/Courseplay_FS22/files/7777638/log.zip)

**Comment by dansiegel1109** [comment]
On a slightly different topic, I found another issue with the Square Baler - CP is not unfolding it before starting.  If the implement is folded, it will go to the course, and sit there, doing nothing - it won't activate the tool, however it does lower the pickup header.  If I dismiss worker, manually unfold, and then hire CP worker, it functions correctly thereafter.  

Again, unique to the square baler compared to round balers which don't t have this problem, the square baler MUST be unfolded to function, however the round balers have nothing to unfold, and therefore work perfectly.

Should I open another ticket and cut/paste the same logs in there?  I reproduced the unfolding issue on all 3 square balers.

**Comment by pvaiko** [comment]
No need to open another ticket, thanks, wish all issues were created so clear and concise like this.

**Comment by pvaiko** [comment]
Not so sure about this stopping/no stopping. Unfortunately I don't know what the Vicon is, I don't have that in the shop, but Giants provided a `nonStopBaling` attribute for balers, apparently if this is set, the balers are not supposed to stop. It is _not_ set for the Krone. 

So for now, I'll leave that as it is. Will still try to fix the unfolding issue though.

**Comment by dansiegel1109** [comment]
Lol, I understand - I'm a retired developer - I know exactly what you mean.  But its worth it for you guys because of the great feedback through Github.  If only Giants did this, they would get more tickets (because people are often frustrated at opening tickets due to no or no apparent feedback/support).

I apologize, the Vicon is technically a mod, however official Giants release, not 3rd Party.  Its a round baler introduced in the Kvernland DLC in FS19, wildly popular because it was the first and only round baler which did not need to stop (it would be VERY slow however if the swath was large, still limited to defined pickup throughput).  Giants released this for free, several days after launch.  Its on modhub and fully supported of course, as its created by Giants.

the nonStopBaling is probably for the Vicon, again as I stated above the Square Baler is a more archaic version of fast baling (non-stopping) which has been around forever.  So, I'm not surprised that nonStopBaling is not set for the Krone or other Square Balers.  

I'll check out the parameter and see if I can modify and test it myself, to see if this fixes the behavior.  However, that may not be a good solution for CP, as I am certain if you ignored this parameter, the classic round balers would foul up and not stop - and they must stop.  Perhaps you can specify in your code that simply Square Balers do not stop.   In gameplay, no square baler (including the tiny tiny square baler) needs to stop to unload.  Its primarily the reason square balers are so expensive and HP requiring, compared to round balers - and why the Vicon Fast Baler is the most popular baler (I'm glad I tested it) - both cheap and non-stopping (just not great for Silage IMO because it slows down to half speed, Giants limitation of grass/silage pickup, although Hay/Straw is at full speed).

thanks for the quick and awesome support!

**Comment by Tensuko** [comment]
Fixed in 7.0.1.0
